### PR TITLE
Commitlog: refactor file handling - simplify file management + make bookkeep safer

### PR DESCRIPTION
**Pull request by elcallio**
Adds a "named_file" wrapper type in commitlog, encapsulating file and disk size, the latter being updated automatically on write/truncate/allocate/delete operations. Use this instead of loose vars in segments, and also in recycle/delete lists.

Having the data propagate with the objects means we can dispose of re-reading sizes from disk, which in turn means we know what "our" view of the file sizes is when we try to delete/recycle them -> we can bookkeep accurately (from our view point) without having to resort to the rather horrible recalculation of disk footprint. 

This series also drops non-recycled segment handling, since it is not used anywhere, and just makes things harder. 
It also adds a parameter to set flush threshold. 
These two first patches could be broken out into separate PR:s if need be.

### CODE REVIEW

File: db/config.cc
Comment (avikivity) [inline db/config.cc:428]: Let's keep the variable (but with value_status::Unused) to avoid problems during upgrade.

File: db/hints/manager.cc
Comment (avikivity) [inline db/hints/manager.cc:397]: What happens during hints replay? Will the segments be converted into recycled segments instead of being deleted?

File: db/config.cc
Comment (avikivity) [inline db/config.cc:428]: Description might hint about algorithm for default.

File: db/commitlog/commitlog.cc
Comment (avikivity) [inline db/commitlog/commitlog.cc:114]: If we make this a parameter, let's listen to it and make it liveupdate so we don't need to restart.

File: db/commitlog/commitlog.cc
Comment (avikivity) [inline db/commitlog/commitlog.cc]: I don't think this is needed. The manual says:

       O_CREAT
              If pathname does not exist, create it as a regular file.

Since it exists, it won't be created. We don't specify O_TRUNC, do we?

File: db/commitlog/commitlog.cc
Comment (avikivity) [inline db/commitlog/commitlog.cc:483]: This just adds an allocation for every write.

File: db/commitlog/commitlog.cc
Comment (avikivity) [inline db/commitlog/commitlog.cc:844]: I appreciate the better consistency, but this is adding an allocation+task to the hot path. It's not quite per write, but per (small number of writes), for the more durable tables.

Comment (avikivity) [reply db/commitlog/commitlog.cc:844]: (per write == per user INSERT/UPDATE, not per commitlog write)

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc]: You are right, but that means the bug is a few lines up, where we treat o_creat as if it where o_creat|o_trunc, and don't read file size in the creat case. Should not ever reach a case where it matters though, esp. if removing non-recycling case.

File: db/hints/manager.cc
Comment (elcallio) [reply db/hints/manager.cc:397]: Yes, up to the disk limit.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:483]: Better than making it a coroutine frame methinks...

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:844]: It is not really the "hot" path... And write is already nestled with continuations (less in refactor coming branch, but...). Seems to me that is is not that hefty an addition.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:114]: Can we do this an a coming exercise? This is converted to a commitlog::config object, passed by value. Doing a changeable thing here will make for very ugly code (not the least because config values are per node, commitlog ones are per shard).

File: db/config.cc
Comment (elcallio) [reply db/config.cc:428]: Done.

**Comment by elcallio** [comment]
Ping? Comments addressed....

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:114]: updateable_value was supposed to address exactly that problem, unfortunately it has a cross-shard bug. Okay to defer it.

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:486]: Sure, but worse than keeping it in the caller.

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:844]: Let's see perf_simple_query with commitlog in batch mode (on ramdisk).

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:114]: Since the configs are separate, the best solution (methinks) would be to add observers in database, and simply re-apply the full  config on change (or rather the parts that can be changed). Still a little over-ugly for this patch set. 
Could also create two different initializers for CL, one with db::config, one with local (for hints).

File: db/config.cc
Comment (elcallio) [reply db/config.cc]: Added a sentence. Not sure how helpful.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:486]: If it really shows up on perf runs (esp. after other rewrites of write path), it can be changed to return a functor object (non-allocating) that simply increments the size on destructor, so `co_await file.dma_write()(pos, buffer, len)` instead. But seems a bit cumbersome unless proven to really matter. Will do the perf runs again.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:844]: For some reason I cannot run `perf_simple_query` on ramdisk.... Keep getting `EINVAL`...
I also had to modify the benchmark to add a commitlog batch option...

Anyway, with plain old tmpfs, we get:
master:
```
	{
		"allocs_per_op" : 68.29561511944604,
		"instructions_per_op" : 0.0,
		"mad tps" : 2118.3316244413145,
		"max tps" : 425416.49492983491,
		"median tps" : 397582.47153378825,
		"min tps" : 395464.13990934694,
		"tasks_per_op" : 18.731985105677328
	},
	{
		"allocs_per_op" : 68.259809725211042,
		"instructions_per_op" : 0.0,
		"mad tps" : 13631.479232312704,
		"max tps" : 420582.24925546983,
		"median tps" : 402449.35569910001,
		"min tps" : 385442.43090700504,
		"tasks_per_op" : 18.712286946097084
	},
	{
		"allocs_per_op" : 68.289302249156691,
		"instructions_per_op" : 0.0,
		"mad tps" : 3096.7571222148254,
		"max tps" : 453216.03738224204,
		"median tps" : 413443.41797986667,
		"min tps" : 397973.08724443399,
		"tasks_per_op" : 18.730835773866481
	},

```

PR:
```
	{
		"allocs_per_op" : 68.385908071775702,
		"instructions_per_op" : 0.0,
		"mad tps" : 9420.6814274010831,
		"max tps" : 466017.08227707446,
		"median tps" : 417543.57998778007,
		"min tps" : 396314.56719759374,
		"tasks_per_op" : 18.80612711258085
	},
	{
		"allocs_per_op" : 68.090368586471001,
		"instructions_per_op" : 0.0,
		"mad tps" : 18115.507142089424,
		"max tps" : 465545.4676978676,
		"median tps" : 425675.20211426448,
		"min tps" : 402974.35145419918,
		"tasks_per_op" : 18.621874361845208
	},
	{
		"allocs_per_op" : 68.119316276455564,
		"instructions_per_op" : 0.0,
		"mad tps" : 24642.845125292952,
		"max tps" : 456051.16822472482,
		"median tps" : 422974.69748250721,
		"min tps" : 388181.71933821432,
		"tasks_per_op" : 18.638215970867595
	},
```

No IPO, since all was run inside docker unfortunately. Not very super stable, but allocs and tasks per op seems to be pretty en par between the two.

**Comment by elcallio** [comment]
More ping...

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:267]: Since this class is internal to commitlog.cc, I shouldn't hold it to a very high standard of documentation, so it's not a disaster that there is none. I just have two minor comments you can feel free not to do:
1. The name "named_file" suggests that tracking the name is the main goal of this file, while it seems its real raison d'etre is tracking the file's size. So maybe the name "size" should have been part of the class name?
2. The method name "known_size()" suggests to me that maybe it doesn't really track the file's size - it's only what is "known" and maybe there're cases when we don't know? What cases are those? If the size is always tracked correctly, and always known, why not just call it size()?

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:289]: Shouldn't this do something about _known_size?

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:292]: When does it make sense to call this function?

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:463]: This assert suggests that instead of a default constructor and one open() call you must do later, you can have a constructor which takes the open parameters, or alternatively , a private constructor and an open() factory (static) function - to get compile-time sanity checking instead of this runtime assert crashes.
But maybe I'm missing something in the big picture that led you to this design?

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:468]: We also have open_flags::truncate, which is spearate from open_flags::create although they are often used together. So shouldn't you have checked open_flags::truncate here?

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:480]: nitpick:
_name = std::move(s)

File: db/commitlog/commitlog.cc
Comment (nyh) [reply db/commitlog/commitlog.cc:486]: I have maybe a stupid question:
Why is it important to update the size only **after** the data was written? What do we lose if we update it immediately? If we update it immediately, even before calling `dma_write()`, we no longer have the extra allocation.

File: db/commitlog/commitlog.cc
Comment (nyh) [inline db/commitlog/commitlog.cc:1580]: Again, I'm not sure if the right thing you wanted to filter was truncate, not create?

**Review by nyh — COMMENTED** [review]
Looks good. I see you are still waiting for Avi's approval (he had some comments regarding the performance), so I took the liberty to add a few comments of my own - most of them trivial but perhaps a couple are worth your attention.

File: db/config.cc
Comment (elcallio) [reply db/config.cc:431]: Segment size is controlled by `commitlog_segment_size_in_mb`. The parameter is per node, but the default is calculated to be roughly "<shard max disk footprint>- <one segment>". Avi wanted a mention of defaults, but they don't exactly roll of the tongue...

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:267]: Actually, carrying the name is probably "primary", since we only care about the size iff we are forced to delete things. But naming is ugh. 
It is called "known size" because it is the size we infer/assume based on partially actually reading size (on open) and IO ops (write, truncate, allocate). Actual file size could be foobar:ed by an external agent, but we ignore this and only care about size we created. So `known` is correct, but only guaranteed so from "our" standpoint. 

(also to not collide with the actual `future<uint64_t> size() const` operation).

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:289]: When we do this we are assuming size will be updated elsewhere or is already correct (only used in allocate_ex file wrapping and named_file::open(...). The type is not 100% self-contained in this. Should maybe make the `assign` take a size parameter as well, so caller must provide it.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:292]: As it is now: Nowhere. Maybe if one did an "assign" from a different call path than stated above. Can remove.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:463]: Recycling. We keep a bunch of named_files in a list, pick one to re-use. These are closed (the seastar::file part is empty). We rename using stored file path, then re-open the object. Could of course just create new objects, using a static open etc. But re-using the object is the motivator here. Assert just means we must call close before calling open again.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:468]: If this was a top-level type, yes. In this case we adapt to the usage within commitlog, where "create" always means new (in the control flow).

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:1580]: In this case it is create actually, because "create" here always means new file (not recycled).

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:486]: If we get an exception, we will have updated it to something that is wrong. Could maybe subtract on exception, but that is just as much an attached task as a then...

Comment (nyh) [reply db/commitlog/commitlog.cc:486]: If we get an exception, does the file size even matter any more? Doesn't it mean the entire file is ruined anyway?

File: db/config.cc
Comment (nyh) [reply db/config.cc:431]: Ok, so if I understand you now, the sentence "Default is per-shard size minus approximately one segment size." is trying to explain what the default is. But what is it saying? What is "per-shard size" referring to? Is this another parameter defined somewhere? Maybe use those parameter names explicitly instead of English that a stupid read (like myself) won't know what they refer to?

File: db/commitlog/commitlog.cc
Comment (nyh) [reply db/commitlog/commitlog.cc:289]: Ok, since it's just an internal utility class, I guess it doesn't need to be 100% beautiful. So ok.

File: db/commitlog/commitlog.cc
Comment (nyh) [reply db/commitlog/commitlog.cc:468]: But don't you also have open_flags::truncate in the same places, so can check that flag instead?
Anyway, I just thought that if you used open_flags::truncate, in addition to (hopefully) being more correct or general, you could get rid of that comment where you needed to apologize for using the wrong flag.

File: db/config.cc
Comment (elcallio) [reply db/config.cc]: I've updated it to just use the config parameter names. Maybe more parse-able now.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:466]: Perhaps. I guess the idea (somewhere) was that iff a file actually existed, we would not truncate it, since doing zeroing etc (that we do) will be faster if disk blocks already are there, but we still define the file size by to where we allocate/truncate/zero-write. Still, a teeny file size read in these cases as well is not the end of the world.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:483]: The file format is built to be somewhat resilient, so no, one failed write is not necessarily the end of the world, though in periodic mode we have no way of back-propagating the info that we mutation write was not backed by a successful CL write. The next buffer write could succeed, and that part of the file at least will be ok. 

Not sure what the best approach is here. My OCD wants to update after the operation. Otoh, we do _pre_ updates for things like write position (because a failure is not terminal). So maybe writes could pre-op update size.

Comment (elcallio) [reply db/commitlog/commitlog.cc:483]: But there is one big caveat: dma_write is not 100% guaranteed to fulfill the whole write op in one go IIRC. That is literally why the write path has a big loop popping off its buffer front. Does not happen often, but at least in the olden days I have seen more than one incomplete write.

Comment (nyh) [reply db/commitlog/commitlog.cc:483]: ׁHmm, so if the caller is anyway forced to call this function in a loop, using continuations, it could update the size.
But I'm out of my element here. I'm not familiar with all those code paths. I'll leave it to you to decide (and to Avi to decide if this extra continuation bothers him).

File: db/config.cc
Comment (avikivity) [reply db/config.cc]: It should be based on per-node limit, people don't normalize to shards. e.g. if system RAM is 100GB, commitlog total space should be 100GB, and the flush threshold be 100GB-nr_shards*segment_size. I'd describe it as "just under the commitlog space limit"; if we have 100 shards and 1GB/shard, it doesn't matter to the user if the limit is 1024 MiB or 992 MiB.

Separately, we need to adjust it to be half the allocated space (per-node when exposed to the user, per-shard internally). We have no reason to expect that we can flush a commitlog's worth of data in the time it takes to fill one segment.

Later we can try to make it smarter by considering the actual used space of the memtables. Something like (1 - (largest memtables size) / (space allocated for memtables)) * (total commitlog space). Note it evaluates to 1/2 when we have one table, since the replica will allow a memtable to at most half the memory allocated to memtables.

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:844]: I did mean tmpfs (do ramdisks still exist? what did you use?)

Looks like your docker ate the performance monitor support so we don't see instructions per op. But tasks and allocations look reasonable.

Comment (elcallio) [reply db/commitlog/commitlog.cc:844]: ramfs vs. tmpfs. And yes, docker kills PM:s, tried to pass it through before but never gotten it to work 100%. 
In any case, even in batch, there are so many continuations involved that one more should make very little difference. And this PR is also meant to lay some groundwork for bigger refactor that tries to slim said write path a bit...

File: db/config.cc
Comment (elcallio) [reply db/config.cc]: Per @nyh:s request I changed the text to "Default is (approximately) commitlog_total_space_in_mb - <num shards>*commitlog_segment_size_in_mb". Which is what I meant and what is (roughly) is. (Some rounding and taking to small total sizes into account for tests).

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:1577]: But O_CREATE is not exclusive. It won't truncate or do anything to the file if it already exists.

Comment (elcallio) [reply db/commitlog/commitlog.cc:1577]: No, in that case we do manual truncation/fill to get to the right size (segment max size). Flipping the bit here is just to save some work for named_file::open. But we could really dispense with that too.

Comment (elcallio) [reply db/commitlog/commitlog.cc]: Changed this to be more consistent. Caller tells size, or we truncate, or we do file::size().

**Review by nyh — APPROVED** [review]
Looks ok to me - but Avi had some performance complaints that don't seem to have been settled, so I'll wait for him to conclude the review and commit this series.

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:483]: It's safer to update post-write, even though it costs a continuation.

**Comment by avikivity** [comment]
Dequeued, commitlog and hinted handoff dtests fail.

**Comment by elcallio** [comment]
Jenkins link?

**Comment by elcallio** [comment]
See /scylladb/scylla-dtest#2632. Merged now.

**Comment by elcallio** [comment]
ping?

**Comment by elcallio** [comment]
ping^2?

**Comment by elcallio** [comment]
Rebased + pushed.

**Comment by benipeled** [comment]
> @benipeled will CI notice an old pull request on a force-push?

Yes - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/87/

**Pull request closed without merging**
